DPL: keep dropping data while in ready, unless a new state was requested

ktf opened the pull request:
DPL: keep dropping data while in ready, unless a new state was requested

For some reason, before we were dropping data only for 5 seconds.
To be seen if this was needed to prevent some timeout by ECS / ODC.

This has now changed to keep dropping data until the "cleanup" property
changes or until there is a NewStatePending().

ktf commented:
I am not sure this is neither what we want, nor what the actual problem is. However, clearly at the moment we stop dropping messages in Ready after 5 seconds from the RUN->READY transition.

alibuild commented:
Error while checking build/O2/fullCI for 2d9ad4c6f4a76b3c3b3f7b8f1be039b17cc337f0 at 2023-11-10 17:48:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/12232-slc8_x86-64/0/Framework/Core/include/Framework/DeviceState.h:62:24: error: field 'cleanupCount' has incomplete type 'std::atomic<long int>'
/sw/SOURCES/O2/12232-slc8_x86-64/0/Framework/Core/include/Framework/DeviceState.h:62:24: error: field 'cleanupCount' has incomplete type 'std::atomic<long int>'
/sw/SOURCES/O2/12232-slc8_x86-64/0/Framework/Core/include/Framework/DeviceState.h:62:24: error: field 'cleanupCount' has incomplete type 'std::atomic<long int>'
/sw/SOURCES/O2/12232-slc8_x86-64/0/Framework/Core/include/Framework/DeviceState.h:62:24: error: field 'cleanupCount' has incomplete type 'std::atomic<long int>'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12232/2d9ad4c6f4a76b3c3b3f7b8f1be039b17cc337f0/build_O2_fullCI/pretty.html).

ktf committed:
Follow coding conventions

ktf committed:
DPL: keep dropping data while in ready, unless a new state was requested

For some reason, before we were dropping data only for 5 seconds.
To be seen if this was needed to prevent some timeout by ECS / ODC.

This has now changed to keep dropping data until the "cleanup" property
changes or until there is a NewStatePending().

davidrohr commented on Framework/Core/src/ExternalFairMQDeviceProxy.cxx:
```diff
@@ -687,53 +687,25 @@ DataProcessorSpec specifyExternalFairMQDeviceProxy(char const* name,
+      // We use it to detect when the device is cleaned up.
+      int64_t cleanupCount = deviceState.cleanupCount.load();
+
+      // Continue iterating we saw the cleanup property being reset or
```
I think there is an `until` missing in the comment

davidrohr reviewed (approved):
Look ok, just a typo in the comment that can be fixed with any later PR...
I'll merge, and build for the EPNs, and test it. I want to verify that the STOP works correctly despite we delay the callback. Worst case we have to revert.

davidrohr commented:
Ahh, wait, QC tests are consistently failing. Could it be that this is because we do not acknowledge the state transision and do not reach READY?

davidrohr commented:
hm, though, I see similar QC ctest timeouts also in other PR. Perhaps let's wait for the CI to rerun, and meanwhile I'll build this one manually without merging

ktf commented:
No, indeed, QC tests die also in other PRs: the FairMQ one you mentioned and also the gRPC one, which only has some cosmetic fixes in the recipe. The macOS errors seems due to some update (@TimoWilken: I do not recall, did we bump some mac builders to Sonoma to check how it fares?)

TimoWilken commented:
@ktf, all the Mac builders are still on Ventura and Xcode 14.3.1, including the two that tested this PR, so I'm not sure what's going on with the `PREALLOCATE` errors.

davidrohr commented:
ok, tested this at P2, seems ok, devices are reaching READY
